gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2

```diff
@@ -36,7 +36,7 @@
             {% elif not method.server_streaming %}
             ) -> {{ method.client_output.ident }}:
             {% else %}
-            ) -> Iterable[{{ method.client_output.ident }}]:
+            ) -> Union[Iterable[{{ method.client_output.ident }}], GrpcStream[{{ method.client_output.ident }}]]:
```

Comment:
This should be redundant, since `GrpcStream` is [just an alias](https://github.com/googleapis/python-api-core/blob/e298ebdfbbb839d64c2da570b454c6c0f6c4e00c/google/api_core/grpc_helpers.py#L149) for `_StreamingResponseIterator`, which is [already Iterable](https://github.com/googleapis/python-api-core/blob/e298ebdfbbb839d64c2da570b454c6c0f6c4e00c/google/api_core/grpc_helpers.py#L101). So we should just be able to use `GrpcStream` here

returning a union type can be ugly, since any attempt to use the returned instance will cause typing issues, since mypy won't know which of the types it is

Reply:
For REST client, `GrpcStream` would be the response type. My initial thought is that we a type which doesn't have `Grpc` in the name. Perhaps we could use a new `StreamResponse` instead of `GrpcStream`?

Reply:
Given that this response type is also used for REST streaming, I'm not sure that we can avoid the union here. `_StreamingResponseIterator` has a dependency on gRPC. We would need to update `_StreamingResponseIterator` if we want to use it for REST

Reply:
As an alternative, we could add a callback similar to what we have for unary requests in https://github.com/googleapis/gapic-generator-python/pull/2243 as a callback could be used to solve the feature request in https://github.com/googleapis/gapic-generator-python/pull/2256#discussion_r1851154197

Reply:
That's a good point, I wasn't considering the rest side. We could make a `StreamResponse` as a mixin class, that we could add to both `_StreamingResponseIterator` and `rest_streaming.ResponseIterator`?

Yeah, using a callback in all cases could be a good way to resolve some of this complexity. It seems like we'd have some of the same typing issues there though, where we still need a way to represent results for either grpc or rest, right? I might have to dig deeper into the rest side